File: apps/mobile/src/db/schema.ts
import { index, real, sqliteTable, text, uniqueIndex } from 'drizzle-orm/sqlite-core';

/**
 * Knowledge item type enum
 * - 'note': User-created text notes
 * - 'link': Saved URLs with optional notes
 * - 'highlight': Key passages excerpted from reading
 * - 'screenshot': Visual captures with optional OCR
 * - 'share': Content received via OS share sheet
 */
export const knowledgeItemType = ['note', 'link', 'highlight', 'screenshot', 'share'] as const;
export type KnowledgeItemType = (typeof knowledgeItemType)[number];

export const knowledgeItemLabelStatus = [
  'idle',
  'pending',
  'provisional',
  'final',
  'failed',
] as const;
export type KnowledgeItemLabelStatus = (typeof knowledgeItemLabelStatus)[number];

export const knowledgeItemLabelSource = [
  'none',
  'rules',
  'apple',
  'local_small',
  'local_full',
  'byok',
] as const;
export type KnowledgeItemLabelSource = (typeof knowledgeItemLabelSource)[number];

/**
 * Knowledge Items table schema
 * Central storage for all user-captured knowledge (notes and links)
 */
export const knowledgeItems = sqliteTable(
  'knowledge_items',
  {
    // Unique identifier (cuid or uuid generated at application layer)
    id: text('id').primaryKey(),

    // Item type: 'note' or 'link'
    type: text('type', { enum: knowledgeItemType }).notNull(),

    // Optional title for the item
    title: text('title'),

    // Main content - required for notes, optional annotation for links
    body: text('body'),

    // URL - required for links, null for notes
    url: text('url'),

    // AI-generated summary (to be filled later by AI processing)
    summary: text('summary'),

    // Tags stored as JSON array string - parsed at application layer
    tags: text('tags', { mode: 'json' }).$type<string[]>(),

    // Lightweight taxonomy labels for archive classification
    labels: text('labels', { mode: 'json' }).$type<string[]>(),
    provisionalLabels: text('provisional_labels', { mode: 'json' }).$type<string[]>(),
    labelStatus: text('label_status', { enum: knowledgeItemLabelStatus }).default('idle'),
    labelSource: text('label_source', { enum: knowledgeItemLabelSource }).default('none'),
    labelVersion: text('label_version'),
    labelScore: real('label_score'),
    labelRequestedAt: real('label_requested_at'),
    labelCompletedAt: real('label_completed_at'),
    labelError: text('label_error'),

    // Creation timestamp (Unix epoch in milliseconds)
    createdAt: real('created_at').notNull(),

    // Last update timestamp (Unix epoch in milliseconds)
    updatedAt: real('updated_at').notNull(),

    // --- Spaced Repetition Fields (MVP v2) ---

    // Memory stability for FSRS algorithm (null if never reviewed)
    stability: real('stability'),

    // Difficulty level for FSRS algorithm (null if never reviewed)
    difficulty: real('difficulty'),

    // Last review timestamp (null if never reviewed)
    lastReviewedAt: real('last_reviewed_at'),

    // Next scheduled review timestamp (null if no review scheduled)
    nextReviewAt: real('next_review_at'),
  },
  (table) => ({
    typeIdx: index('knowledge_items_type_idx').on(table.type),
    createdAtIdx: index('knowledge_items_created_at_idx').on(table.createdAt),
    nextReviewAtIdx: index('knowledge_items_next_review_at_idx').on(table.nextReviewAt),
    labelStatusIdx: index('knowledge_items_label_status_idx').on(table.labelStatus),
    labelRequestedAtIdx: index('knowledge_items_label_requested_at_idx').on(table.labelRequestedAt),
  })
);

type KnowledgeItemRow = typeof knowledgeItems.$inferSelect;
type NewKnowledgeItemRow = typeof knowledgeItems.$inferInsert;

type OptionalKnowledgeItemLabelFields =
  | 'labels'
  | 'provisionalLabels'
  | 'labelStatus'
  | 'labelSource'
  | 'labelVersion'
  | 'labelScore'
  | 'labelRequestedAt'
  | 'labelCompletedAt'
  | 'labelError';

// Keep new labeling fields optional at the type boundary while the feature rolls out.
export type KnowledgeItem = Omit<KnowledgeItemRow, OptionalKnowledgeItemLabelFields> &
  Partial<Pick<KnowledgeItemRow, OptionalKnowledgeItemLabelFields>>;
export type NewKnowledgeItem = Omit<NewKnowledgeItemRow, OptionalKnowledgeItemLabelFields> &
  Partial<Pick<NewKnowledgeItemRow, OptionalKnowledgeItemLabelFields>>;

/**
 * Recommendation status enum
 * - 'pending': Initial state, awaiting user action
 * - 'accepted': User explicitly accepted the connection
 * - 'ignored': User explicitly rejected
 * - 'dismissed': User temporarily hid (may resurface later)
 */
export const recommendationStatus = ['pending', 'accepted', 'ignored', 'dismissed'] as const;
export type RecommendationStatus = (typeof recommendationStatus)[number];

/**
 * Recommendations table schema
 * Stores AI-suggested connections between knowledge items
 */
export const recommendations = sqliteTable(
  'recommendations',
  {
    // Unique identifier
    id: text('id').primaryKey(),

    // First item in the connection
    itemA_id: text('item_a_id')
      .notNull()
      .references(() => knowledgeItems.id),

    // Second item in the connection
    itemB_id: text('item_b_id')
      .notNull()
      .references(() => knowledgeItems.id),

    // Reason for the recommendation
    reason: text('reason'),

    // Current status
    status: text('status', { enum: recommendationStatus }).notNull().default('pending'),

    // Creation timestamp
    createdAt: real('created_at').notNull(),

    // Response timestamp (when user accepted/ignored/dismissed)
    respondedAt: real('responded_at'),
  },
  (table) => ({
    statusIdx: index('recommendations_status_idx').on(table.status),
    itemAIdx: index('recommendations_item_a_idx').on(table.itemA_id),
    itemBIdx: index('recommendations_item_b_idx').on(table.itemB_id),
    pairUniqueIdx: uniqueIndex('recommendations_item_pair_unique_idx').on(
      table.itemA_id,
      table.itemB_id
    ),
  })
);

// Type exports for recommendations
export type Recommendation = typeof recommendations.$inferSelect;
export type NewRecommendation = typeof recommendations.$inferInsert;

/**
 * Feedback action type enum
 * - 'accept': User accepted the recommendation
 * - 'ignore': User explicitly rejected
 * - 'dismiss': User temporarily dismissed
 */
export const feedbackActionType = ['accept', 'ignore', 'dismiss'] as const;
export type FeedbackActionType = (typeof feedbackActionType)[number];

/**
 * Feedback Events table schema
 * Logs user reactions to recommendations for learning/analytics
 */
export const feedbackEvents = sqliteTable(
  'feedback_events',
  {
    // Unique identifier
    id: text('id').primaryKey(),

    // Reference to the recommendation
    recommendationId: text('recommendation_id')
      .notNull()
      .references(() => recommendations.id),

    // Action taken by user
    action: text('action', { enum: feedbackActionType }).notNull(),

    // Timestamp of the event
    createdAt: real('created_at').notNull(),
  },
  (table) => ({
    recommendationIdIdx: index('feedback_events_recommendation_id_idx').on(
      table.recommendationId
    ),
    createdAtIdx: index('feedback_events_created_at_idx').on(table.createdAt),
  })
);

// Type exports for feedback events
export type FeedbackEvent = typeof feedbackEvents.$inferSelect;
export type NewFeedbackEvent = typeof feedbackEvents.$inferInsert;

/**
 * Message role enum
 * - 'user': Message from the user
 * - 'assistant': Message from the AI
 */
export const messageRole = ['user', 'assistant'] as const;
export type MessageRole = (typeof messageRole)[number];

/**
 * Conversations table schema
 * Stores chat conversations with optional context item
 */
export const conversations = sqliteTable(
  'conversations',
  {
    // Unique identifier
    id: text('id').primaryKey(),

    // Conversation title (auto-generated or user-set)
    title: text('title'),

    // Optional icon selected by the user
    icon: text('icon'),

    // Optional context item ID (from library)
    contextItemId: text('context_item_id'),

    // Creation timestamp
    createdAt: real('created_at').notNull(),

    // Last update timestamp
    updatedAt: real('updated_at').notNull(),

    // Soft delete timestamp
    deletedAt: real('deleted_at'),
  },
  (table) => ({
    createdAtIdx: index('conversations_created_at_idx').on(table.createdAt),
    contextItemIdx: index('conversations_context_item_idx').on(table.contextItemId),
  })
);

// Type exports for conversations
export type Conversation = typeof conversations.$inferSelect;
export type NewConversation = typeof conversations.$inferInsert;

/**
 * Messages table schema
 * Stores individual messages within conversations
 */
export const messages = sqliteTable(
  'messages',
  {
    // Unique identifier
    id: text('id').primaryKey(),

    // Reference to conversation
    conversationId: text('conversation_id')
      .notNull()
      .references(() => conversations.id),

    // Message role (user or assistant)
    role: text('role', { enum: messageRole }).notNull(),

    // Message content
    content: text('content').notNull(),

    // Creation timestamp
    createdAt: real('created_at').notNull(),

    // Last update timestamp (null if never edited)
    updatedAt: real('updated_at'),

    // Soft delete timestamp (null if not deleted)
    deletedAt: real('deleted_at'),
  },
  (table) => ({
    conversationIdx: index('messages_conversation_idx').on(table.conversationId),
    createdAtIdx: index('messages_created_at_idx').on(table.createdAt),
  })
);

// Type exports for messages
export type Message = typeof messages.$inferSelect;
export type NewMessage = typeof messages.$inferInsert;

/**
 * Embedding source type enum
 * - 'message': Embedding of a chat message
 * - 'knowledge_item': Embedding of a library item
 */
export const embeddingSourceType = ['message', 'knowledge_item'] as const;
export type EmbeddingSourceType = (typeof embeddingSourceType)[number];

/**
 * Embeddings table schema
 * Stores vector embeddings for RAG (Retrieval Augmented Generation)
 */
export const embeddings = sqliteTable(
  'embeddings',
  {
    // Unique identifier
    id: text('id').primaryKey(),

    // Source type
    sourceType: text('source_type', { enum: embeddingSourceType }).notNull(),

    // Source ID (message or knowledge item)
    sourceId: text('source_id').notNull(),

    // Vector embedding stored as JSON array
    vector: text('vector', { mode: 'json' }).$type<number[]>().notNull(),

    // Creation timestamp
    createdAt: real('created_at').notNull(),
  },
  (table) => ({
    sourceTypeIdx: index('embeddings_source_type_idx').on(table.sourceType),
    sourceIdIdx: index('embeddings_source_id_idx').on(table.sourceId),
  })
);

// Type exports for embeddings
export type Embedding = typeof embeddings.$inferSelect;
export type NewEmbedding = typeof embeddings.$inferInsert;

File: packages/core-rs/src/db/schema.rs
use std::path::Path;

use anyhow::Result;
use rusqlite::Connection;

use super::patches::normalize_json_string_array;

pub const DB_NAME: &str = "glimpse.db";

pub const CREATE_KNOWLEDGE_ITEMS_TABLE_SQL: &str = r#"
CREATE TABLE IF NOT EXISTS knowledge_items (
  id TEXT PRIMARY KEY NOT NULL,
  type TEXT NOT NULL CHECK(type IN ('note', 'link', 'highlight', 'screenshot', 'share')),
  title TEXT,
  body TEXT,
  url TEXT,
  summary TEXT,
  tags TEXT,
  labels TEXT,
  provisional_labels TEXT,
  label_status TEXT DEFAULT 'idle' CHECK(label_status IN ('idle', 'pending', 'provisional', 'final', 'failed')),
  label_source TEXT DEFAULT 'none' CHECK(label_source IN ('none', 'rules', 'apple', 'local_small', 'local_full', 'byok')),
  label_version TEXT,
  label_score REAL,
  label_requested_at REAL,
  label_completed_at REAL,
  label_error TEXT,
  created_at REAL NOT NULL,
  updated_at REAL NOT NULL,
  stability REAL,
  difficulty REAL,
  last_reviewed_at REAL,
  next_review_at REAL
);
"#;

pub const CREATE_RECOMMENDATIONS_TABLE_SQL: &str = r#"
CREATE TABLE IF NOT EXISTS recommendations (
  id TEXT PRIMARY KEY NOT NULL,
  item_a_id TEXT NOT NULL,
  item_b_id TEXT NOT NULL,
  reason TEXT,
  status TEXT NOT NULL DEFAULT 'pending' CHECK(status IN ('pending', 'accepted', 'ignored', 'dismissed')),
  created_at REAL NOT NULL,
  responded_at REAL,
  FOREIGN KEY (item_a_id) REFERENCES knowledge_items(id),
  FOREIGN KEY (item_b_id) REFERENCES knowledge_items(id)
);
"#;

pub const CREATE_FEEDBACK_EVENTS_TABLE_SQL: &str = r#"
CREATE TABLE IF NOT EXISTS feedback_events (
  id TEXT PRIMARY KEY NOT NULL,
  recommendation_id TEXT NOT NULL,
  action TEXT NOT NULL CHECK(action IN ('accept', 'ignore', 'dismiss')),
  created_at REAL NOT NULL,
  FOREIGN KEY (recommendation_id) REFERENCES recommendations(id)
);
"#;

pub const CREATE_CONVERSATIONS_TABLE_SQL: &str = r#"
CREATE TABLE IF NOT EXISTS conversations (
  id TEXT PRIMARY KEY NOT NULL,
  title TEXT,
  icon TEXT,
  context_item_id TEXT,
  created_at REAL NOT NULL,
  updated_at REAL NOT NULL,
  deleted_at REAL
);
"#;

pub const CREATE_MESSAGES_TABLE_SQL: &str = r#"
CREATE TABLE IF NOT EXISTS messages (
  id TEXT PRIMARY KEY NOT NULL,
  conversation_id TEXT NOT NULL,
  role TEXT NOT NULL CHECK(role IN ('user', 'assistant')),
  content TEXT NOT NULL,
  created_at REAL NOT NULL,
  updated_at REAL,
  deleted_at REAL,
  FOREIGN KEY (conversation_id) REFERENCES conversations(id)
);
"#;

pub const CREATE_EMBEDDINGS_TABLE_SQL: &str = r#"
CREATE TABLE IF NOT EXISTS embeddings (
  id TEXT PRIMARY KEY NOT NULL,
  source_type TEXT NOT NULL CHECK(source_type IN ('message', 'knowledge_item')),
  source_id TEXT NOT NULL,
  vector TEXT NOT NULL,
  created_at REAL NOT NULL
);
"#;

pub const CREATE_INDEXES_SQL: [&str; 17] = [
    "CREATE INDEX IF NOT EXISTS knowledge_items_type_idx ON knowledge_items(type);",
    "CREATE INDEX IF NOT EXISTS knowledge_items_created_at_idx ON knowledge_items(created_at);",
    "CREATE INDEX IF NOT EXISTS knowledge_items_next_review_at_idx ON knowledge_items(next_review_at);",
    "CREATE INDEX IF NOT EXISTS knowledge_items_label_status_idx ON knowledge_items(label_status);",
    "CREATE INDEX IF NOT EXISTS knowledge_items_label_requested_at_idx ON knowledge_items(label_requested_at);",
    "CREATE INDEX IF NOT EXISTS recommendations_status_idx ON recommendations(status);",
    "CREATE INDEX IF NOT EXISTS recommendations_item_a_idx ON recommendations(item_a_id);",
    "CREATE INDEX IF NOT EXISTS recommendations_item_b_idx ON recommendations(item_b_id);",
    "CREATE UNIQUE INDEX IF NOT EXISTS recommendations_item_pair_unique_idx ON recommendations(item_a_id, item_b_id);",
    "CREATE INDEX IF NOT EXISTS feedback_events_recommendation_id_idx ON feedback_events(recommendation_id);",
    "CREATE INDEX IF NOT EXISTS feedback_events_created_at_idx ON feedback_events(created_at);",
    "CREATE INDEX IF NOT EXISTS conversations_created_at_idx ON conversations(created_at);",
    "CREATE INDEX IF NOT EXISTS conversations_context_item_idx ON conversations(context_item_id);",
    "CREATE INDEX IF NOT EXISTS messages_conversation_idx ON messages(conversation_id);",
    "CREATE INDEX IF NOT EXISTS messages_created_at_idx ON messages(created_at);",
    "CREATE INDEX IF NOT EXISTS embeddings_source_type_idx ON embeddings(source_type);",
    "CREATE INDEX IF NOT EXISTS embeddings_source_id_idx ON embeddings(source_id);",
];

pub fn open_in_memory() -> Result<Connection> {
    Ok(Connection::open_in_memory()?)
}

pub fn open_database(data_path: &str) -> Result<Connection> {
    let db_path = Path::new(data_path).join(DB_NAME);
    let conn = Connection::open(db_path)?;
    initialize_schema(&conn)?;
    Ok(conn)
}

pub fn initialize_schema(conn: &Connection) -> Result<()> {
    conn.execute_batch(CREATE_KNOWLEDGE_ITEMS_TABLE_SQL)?;
    conn.execute_batch(CREATE_RECOMMENDATIONS_TABLE_SQL)?;
    conn.execute_batch(CREATE_FEEDBACK_EVENTS_TABLE_SQL)?;
    conn.execute_batch(CREATE_CONVERSATIONS_TABLE_SQL)?;
    conn.execute_batch(CREATE_MESSAGES_TABLE_SQL)?;
    conn.execute_batch(CREATE_EMBEDDINGS_TABLE_SQL)?;
    migrate_knowledge_items_schema(conn)?;

    for statement in CREATE_INDEXES_SQL {
        conn.execute(statement, [])?;
    }

    sanitize_knowledge_item_json_columns(conn)?;

    Ok(())
}

fn migrate_knowledge_items_schema(conn: &Connection) -> Result<()> {
    let existing_columns = get_table_columns(conn, "knowledge_items")?;

    for (column_name, column_definition) in [
        ("labels", "TEXT"),
        ("provisional_labels", "TEXT"),
        (
            "label_status",
            "TEXT DEFAULT 'idle' CHECK(label_status IN ('idle', 'pending', 'provisional', 'final', 'failed'))",
        ),
        (
            "label_source",
            "TEXT DEFAULT 'none' CHECK(label_source IN ('none', 'rules', 'apple', 'local_small', 'local_full', 'byok'))",
        ),
        ("label_version", "TEXT"),
        ("label_score", "REAL"),
        ("label_requested_at", "REAL"),
        ("label_completed_at", "REAL"),
        ("label_error", "TEXT"),
        ("stability", "REAL"),
        ("difficulty", "REAL"),
        ("last_reviewed_at", "REAL"),
        ("next_review_at", "REAL"),
    ] {
        if !existing_columns.iter().any(|existing| existing == column_name) {
            let sql =
                format!("ALTER TABLE knowledge_items ADD COLUMN {column_name} {column_definition};");
            conn.execute(&sql, [])?;
        }
    }

    Ok(())
}

fn get_table_columns(conn: &Connection, table_name: &str) -> Result<Vec<String>> {
    let pragma = format!("PRAGMA table_info({table_name});");
    let mut statement = conn.prepare(&pragma)?;
    let columns = statement
        .query_map([], |row| row.get::<_, String>(1))?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    Ok(columns)
}

pub fn sanitize_knowledge_item_json_columns(conn: &Connection) -> Result<()> {
    let mut statement =
        conn.prepare("SELECT id, tags, labels, provisional_labels FROM knowledge_items;")?;
    let rows = statement.query_map([], |row| {
        Ok((
            row.get::<_, String>(0)?,
            row.get::<_, Option<String>>(1)?,
            row.get::<_, Option<String>>(2)?,
            row.get::<_, Option<String>>(3)?,
        ))
    })?;

    for row in rows {
        let (id, tags, labels, provisional_labels) = row?;
        update_json_column(conn, &id, "tags", normalize_json_string_array(tags))?;
        update_json_column(conn, &id, "labels", normalize_json_string_array(labels))?;
        update_json_column(
            conn,
            &id,
            "provisional_labels",
            normalize_json_string_array(provisional_labels),
        )?;
    }

    Ok(())
}

fn update_json_column(
    conn: &Connection,
    item_id: &str,
    column_name: &str,
    next_value: Option<String>,
) -> Result<()> {
    let sql = format!("UPDATE knowledge_items SET {column_name} = ?1 WHERE id = ?2;");
    conn.execute(&sql, (&next_value, item_id))?;
    Ok(())
}
